org.eclipse.wildwebdeveloper/src/org/eclipse/wildwebdeveloper/vue/autoinsert/VueAutoInsertReconciler.java

```diff
@@ -0,0 +1,209 @@
+/*******************************************************************************
+ * Copyright (c) 2022, 2023 Red Hat Inc. and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Angelo ZERR (Red Hat Inc.) - initial implementation
+ *  Dawid Pakuła - modified copy for Vue
+ *******************************************************************************/
+package org.eclipse.wildwebdeveloper.vue.autoinsert;
+
+import java.net.URI;
+import java.util.concurrent.CompletableFuture;
+
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.DocumentEvent;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.IDocumentListener;
+import org.eclipse.jface.text.ITextInputListener;
+import org.eclipse.jface.text.ITextViewer;
+import org.eclipse.jface.text.reconciler.IReconciler;
+import org.eclipse.jface.text.reconciler.IReconcilingStrategy;
+import org.eclipse.lsp4e.LSPEclipseUtils;
+import org.eclipse.lsp4e.LanguageServers;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
+import org.eclipse.swt.widgets.Display;
+import org.eclipse.wildwebdeveloper.Activator;
+import org.eclipse.wildwebdeveloper.vue.VueLanguageServerAPI;
+
+/**
+ * {@link IReconciler} implementation used to support auto close tags / auto
+ * insert quote, features provides by the vscode VUE language server with the
+ * custom 'volar/client/autoInsert' LSP request.
+ *
+ */
+public class VueAutoInsertReconciler implements IReconciler {
+
+	private IDocument document;
+
+	private ITextViewer viewer;
+
+	private Listener listener;
+
+	private void autoInsert(DocumentEvent event) {
+		if (event == null || viewer == null) {
+			return;
+		}
+		IDocument document = event.getDocument();
+		if (document == null || event == null || event.getLength() != 0) {
+			return;
+		}
+
+		int offset = event.getOffset();
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri == null) {
+			return;
+		}
+		
+		
+		TextDocumentIdentifier identifier = new TextDocumentIdentifier(uri.toString());
+		
+		LanguageServers.forDocument(document).collectAll((w, ls) -> CompletableFuture.completedFuture(ls))
+				.thenAccept(lss -> lss.stream().filter(VueLanguageServerAPI.class::isInstance)
+						.map(VueLanguageServerAPI.class::cast).findAny().ifPresent(info -> {
+							// The document is bound with HTML language server, consumes the html/autoInsert
+							final Display display = viewer.getTextWidget().getDisplay();
+							CompletableFuture.supplyAsync(() -> {
+								try {
+									// Wait for textDocument/didChange
+									Thread.sleep(100);

```

Comment:
This code that I de for XML is so bad and can give some blocking thread bugs.

@mickaelistria LSP4e should provide an API to execute somerhing after à didChange

Reply:
I off course agree, for now this is copy-paste from HTMLAutoInsertReconciler

Reply:
I made some tests and prepared pull request for lsp4e: https://github.com/eclipse/lsp4e/pull/812